_drafts/tracking-code-size.md

```diff
@@ -0,0 +1,598 @@
+---
+title: "Tracking Firmware Code Size"
+description:
+  "Tracking code size for embedded software projects can prevent surprises,
+  maintain productivity, and give insight to developers looking to retroactively
+  reduce a firmware binary size."
+tag: [fw-code-size]
+author: tyler
+image: /img/tracking-code-size/tracking-code-size.png
+---
+
+```
+arm-none-eabi-ld: firmware.elf section `.text' will not fit in region `ROM'
+arm-none-eabi-ld:: region `ROM' overflowed by 16 bytes
+make: error: arm-none-eabi-ld returned 1 exit status
+make: *** [firmware.elf] Error 1
+```
+
+This is the terminal output of my nightmares. It frequently means a halt in
+productivity and sends engineers scrambling to save 100 bytes wherever they can,
+all while trying to meet the deadlines and requirements set forth by the Product
+team.
+
+Fortunately, this message doesn't always have to come as a surprise. Where I've
+worked in the past, we had the tools and infrastructure in place to know when we
+would run out of code space, as well as where it all went. These tools were
+immensely helpful when our team at Pebble was trying to
+[cram an 800kB firmware into a 512kB ROM device](http://web.archive.org/web/20160308073714/https://blog.getpebble.com/2015/12/09/3ontintin/).
+
+Below is an example of a type of dashboard we had. When we approached 5kB
+remaining in our firmware image, we knew what to do.
+
+![](/img/tracking-code-size/running-out-code-space.png)
+
+Hopefully this post provides the tools you need to have a dashboard like this
+too!
+
+<!-- excerpt start -->
+
+Code size information should be surfaced and tracked throughout the development
+of an embedded software project, especially if there is even a slight chance of
+running out in the future. In this post, we'll cover why this is the case, show
+how to track code size, and show how to calculate code size deltas on pull
+requests to help keep code size to a minimum.
+
+<!-- excerpt end -->
+
+_Like Interrupt? [Subscribe](http://eepurl.com/gpRedv) to get our latest posts
+straight to your mailbox_
+
+{:.no_toc}
+
+## Table of Contents
+
+<!-- prettier-ignore -->
+* auto-gen TOC:
+{:toc}
+
+## Why Track Code Size?
+
+If you are a firmware engineer, you've likely run into code sizes issues. If
+not, it is likely because the firmware engineers before you fought for an MCU
```

Comment:
I would remove this second sentence, and merge this into the next paragram. Also remove the s after size.

"If you are a firmware engineer, you've likely run into code size~~s~~ issues. Firmware, like a gas, will expand to take up all available space, whether with extra debug logs, a full featured command line interface, or fancy animation & images."